### ISSUE TITLE
Trouble loading matlab file

### ISSUE THREAD

**Issue by RyanHope** [frejanordsiek/hdf5storage#55]
CPU at 100% but file never loads. Here is a link to the file: https://cmu.box.com/shared/static/gwi0m0uls4rgbf1f5t5zeaq3k2ea2293.set
```
>>> import hdf5storage
>>> import numpy as np
>>> out = hdf5storage.read(path="/EEG",filename="debug.set")
```

**Comment by frejanordsiek** [comment]
I checked the HDF5 file and figured out why it took so long to load. The file has very large HDF5 reference arrays, 211,186 elements. This package is notoriously slow at handling such references, which is why your file fails to load in a reasonable amount of time. The problem is the struct array urevent in EEG, which is 41913 elements long. Each element of each field has to be stored as a separate Dataset in the HDF5 file and pointed to with a reference.

**Comment by horsto** [comment]
I have the same problem here with loading a .mat file (-v7.3). The file is 2.9 GB big and I want to load a Matlab cell array from it (among other variables). Is there anything that can be done to speed up the loading process (or let it finish in the first place)? 

`hdf5storage.loadmat(BIG_MAT_FILE, variable_names=['cellsData', 'allStatistics','p'])
`
(cellsData is the largest variable)

**Comment by frejanordsiek** [comment]
Sadly, there is no trick to speed it up without changing the data. There is a lot of overhead for each cell-array-element/variable. Almost none of the time is taken up by writing or reading actual numerical array elements to/from disk, so a medium sized double array takes hardly more time than a 1-element double array. This is one of the big weaknesses of the package at present. I am working towards the next feature release (at some point) which will hopefully be feature complete, and then after that I am going to focus on the performance issues. Unfortunately, there may not be that much improvement that is possible before hitting the limits of interpreted python (as of right now, the package is a pure python package). There might be some more performance that can be squeezed out by going over to Cython but then has to be built on the computers of people using it, but this might be a worthwhile price to pay. Do you have any thoughts on this eventually going over to Cython?

There is a way to benchmark hdf5storage's performance on your computer and then figure out how long it will take to read it and decide if you want to commit that amount of time or not. The following code writes 10k elements without metadata (set the metadata flags to True to see how they affect it) and then reads it back.

```
>>> import numpy as np
>>> import hdf5storage
>>> a = np.object_([np.int8(1)]*10000)
>>> %time hdf5storage.write(a, path='/a', matlab_compatible=False, store_python_metadata=False)
>>> %time b = hdf5storage.read(path='/a')
```
Back in 2015 I benchmarked hdf5storage for cell arrays on what was then an outdated machine (Core 2 Duo machine that was at from 2010 or earlier) and got a write speed of 172 elements/s (write time of 58 s) and a read speed of 15 elements/s (read time of 665 s).

On my laptop right now (python 3.5.3 on a 4th gen Intel i5 writing and reading to a tmpfs style ramdisk), I got a write speed of 149 elements/s (write time of 67 s) and a read speed of 19 elements/s (read time of 534 s).

**Comment by frejanordsiek** [comment]
I did quite a few small changes that help performance a small amount (see commit 1c4c4c9) on the master branch. It doesn't significantly help the above benchmark, but it does reduce overhead when there is metadata (Attributes) present so it might help a bit. There is a lot more work to do, though.

**Comment by frejanordsiek** [comment]
I managed to vastly improve the read speed by several different optimzations for the master branch (commits listed at the end).

On my laptop, I got a write speed of 207 elements/s (write time of 48 s) and a read speed of **2924 elements/s)** (read time of 3.42 s) for this particular example (10k element cell-array like object with int8 scalars in each element).

This is an improvement of a factor of 154.

Only bad news is that the biggest improvements came from changes that are impossible to extend this to the 0.1.x branch without breaking compatibility (had to change the call signatures ``read`` method of the Marshallers)

The commits are

0a87254
4280fd0
1ebb29d
5e6400a
c93472f

Basically, the code, despite having the Dataset and Group objects and reading directly, it was taking the objects to find the parent group and the object name and then looking up the Dataset or Group object again; along with slow lookups using lists instead of sets and dicts and unnecessary deepcopying of marshallers. Old code and APIs from when I first started this project while being a beginner Python programmer (I suspect there is a lot more performance optimization I can squeeze out by checking all the old code again).

**Comment by frejanordsiek** [comment]
Managed to increase the write performance by a factor of ten. On my laptop, I got 2070 elements/s (write time of 4.83 s)

The commits are 58fbbf4 and 6f14019. Unlike the changes that improved read performance, these could be applied backwards to the 0.1.x branch. The issue turned out to be that when trying to generate random names to put objects being referenced, rather than checking if each random name was in the group or not, all names in the entire group were grabbed first thing and then that list checked, which becomes very slow once there is a non-negligible number of objects in the Group.

**Comment by frejanordsiek** [comment]
Managed to increase the performance a bit more. On my laptop, I got a write speed of 2660 elements/s (write time of 3.72 s) and a read speed of 5350 elements/s (read time of 1.87 s). The total improvement from the first benchmarks is 18 for writing and 280 for reading.

New commits are 7773817, 0531684, and 6abd167 on the master branch.
